modules/openid_connect/db/migrate/20240829140616_migrate_oidc_settings_to_providers.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class MigrateOidcSettingsToProviders < ActiveRecord::Migration[7.1]
+  def up
+    providers = Hash(Setting.plugin_openproject_openid_connect).with_indifferent_access[:providers]
+    return if providers.blank?
+
+    providers.each do |name, configuration|
+      migrate_provider!(name, configuration)
+    end
+  end
+
+  def down
+    # This migration does not yet remove Setting.plugin_openproject_openid_connect
+    # so it can be retried.
+  end
+
+  private
+
+  def migrate_provider!(name, configuration)
+    Rails.logger.debug { "Trying to migrate OpenID provider #{name} from previous settings format..." }
+    call = ::OpenIDConnect::SyncService.new(name, configuration).call
```

Comment:
If we are going to use code in a migration that uses AR models as well, then we should at least add a spec for this migration so that we notice when (not if) it breaks in the future due to changes in the AR models or other code changes.

This is not covered by the ProviderSeeder spec. I mean a spec which sets up the database schema at the time of the migration and checks that migration results in the data in the schema valid at this time as well, via plain SQL. This plain SQL can either be in the migration itself to ensure it works in the future, or we use plain SQL to verify the migration working in the migration spec.

I know it's tedious but could you please have a look?